test/e2e/clusterclass_changes.go

```diff
@@ -300,6 +355,22 @@ func modifyControlPlaneViaClusterClassAndWait(ctx context.Context, input modifyC
 			g.Expect(ok).To(BeTrue(), fmt.Sprintf("failed to get field %q", fieldPath))
 			g.Expect(currentValue).To(Equal(expectedValue), fmt.Sprintf("field %q should be equal", fieldPath))
 		}
+
+		replicas, ok, err := unstructured.NestedFieldCopy(controlPlane.Object, contract.ControlPlane().Replicas().Path()...)
+		g.Expect(err).ToNot(HaveOccurred())
+		g.Expect(ok).To(BeTrue())
+
+		readyReplicas, ok, err := unstructured.NestedFieldCopy(controlPlane.Object, contract.ControlPlane().ReadyReplicas().Path()...)
+		g.Expect(err).ToNot(HaveOccurred())
+		g.Expect(ok).To(BeTrue())
+
+		updatedReplicas, ok, err := unstructured.NestedFieldCopy(controlPlane.Object, contract.ControlPlane().UpdatedReplicas().Path()...)
+		g.Expect(err).ToNot(HaveOccurred())
+		g.Expect(ok).To(BeTrue())
+
+		g.Expect(replicas).To(BeComparableTo(readyReplicas), "Machine deployment does not have expected replicas")
+		g.Expect(replicas).To(BeComparableTo(updatedReplicas), "Machine deployment does not have expected updated replicas")
```

Comment:
```suggestion
		// Ensure KCP recognized the change and finished scaling.
		// Note: This is to ensure to not hit https://github.com/kubernetes-sigs/cluster-api/issues/11772
		observedGeneration, ok, err := unstructured.NestedInt64(controlPlane.Object, "status", "observedGeneration")
		g.Expect(err).ToNot(HaveOccurred())
		if ok {
			g.Expect(controlPlane.GetGeneration()).To(BeComparableTo(observedGeneration))
		}
		scaling, err := contract.ControlPlane().IsScaling(controlPlane)
		g.Expect(err).ToNot(HaveOccurred())
		g.Expect(scaling).To(BeFalse())
```

The old change does not catch everything. This should ensure KCP reconciled the change and finished the rollout.

Note: this should also fix the e2e test issue then reliably for this test which comes down to what is described here https://github.com/kubernetes-sigs/cluster-api/issues/11772